Geometry_Engine/Query/IsNullOrContainsNulls.cs

```diff
@@ -0,0 +1,40 @@
+﻿using BH.oM.Geometry;
+using BH.oM.Reflection.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BH.Engine.Geometry
+{
+    /***************************************************/
+    /**** Public Methods                            ****/
+    /***************************************************/
+    public static partial class Query
+    {
+        [Description("Checks if a List of Geometry is null and if any element in the List is null and outputs relevant error message.")]
```

Comment:
@rolyhudson I added an `IsNullOrEmpty` in the BHoM_Engine here: https://github.com/BHoM/BHoM_Engine/blob/master/BHoM_Engine/Query/IsNullOrEmpty.cs

The thinking was it will check if the `IEnumerable<T>` was null or it was empty, and then within the actual code you would make use of the `IsNull` method for the object itself . `IEnumberable<T>.Any(x => x.IsNull())`

This way the error message returned is not generic and tells you what object is null rather than an object in the list is null.

Reply:
@peterjamesnugent cheers for looking at this appreciate your thoughts. 
I did see that method and the reasons I did not implement as you suggest are multiple...

First up I'm conscious of avoiding adding excessive lines of code with this checking of null handling. The intention of the  `IsNullOrContainsNulls` is to try and manage what could be come a maintenance headache by providing a single method to check any collection of IGeometries.

I am aware of the need for decent message reporting for users. These checks could run quite deep so I am not sure tracking the only object that is null is necessarily helpful. 
For example if a null Point is found in the ControlPoints of one of a an input collection of Polylines, is it useful to report a Point was found to be null?

To be useful would, I think, require providing a stack of messages tracking back to the offending input, which in some situations could be inefficient as the string processing itself could slow things down in some situations.

My thinking (yet to be committed) is just to report the name of the method input parameter where a null was found and nothing more.

It is still a WIP and really needs scrutiny from others before rolling out across the 600+ methods. Maybe worth a conversation with  you very soon.

Reply:
Understand your point about excessive lines of null checks - one thing that did come up on the Structure_Engine PR would be an option for a "deep check", especially for geometry objects that may be worth considering on this PR: https://github.com/BHoM/BHoM_Engine/issues/2487#issuecomment-853314761

> For example if a null Point is found in the ControlPoints of one of a an input collection of Polylines, is it useful to report a Point was found to be null?

With the `[CallerMemberName]` attribute it would feedback that "Cannot evaluate ControlPoints because a Point is null" - but I see your point about traceability, especially given that so many other Engine methods make use of Geometry_Engine methods. The user might not even know the `ControlPoints` method is being used and leads to even more confusion in the error message!

Reply:
@peterjamesnugent - updated this following the conversation. The method `ContainsNulls` cannot move to the base as it is specific to `IEnumerable<IGeometry>` as it call `Geometry.Query.IIsNull()` on each element in the collection. 

This is looking better but I think still needs further tuning in how the messages are handled.

I have added a basic test script  [here](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/BHoM_Engine/Geometry_Engine/%232516-NullHandling?csf=1&web=1&e=rDCZ9p). Quite simple for now - just working with Polylines and the BooleanDifference method.